scripts/update/update_genes.py

```diff
@@ -0,0 +1,571 @@
+#!/usr/bin/env python3
+
+import os.path
+import sys
+import argparse
+import MySQLdb
+import urllib.request
+import configparser
+import gzip
+import re
+
+def locus_id_foreign_key_check(db_host, db_port, db_name, user, password, g2p_tables_with_locus_id_link):
+    db = MySQLdb.connect(host=db_host, port=db_port, user=user, passwd=password, db=db_name)
+    cursor = db.cursor()
+
+    for table in g2p_tables_with_locus_id_link:
+        column_name = "gene_id"
+
+        if table.startswith("locus_") or table.startswith("gene2phenotype_app_"):
+            column_name = "locus_id"
+
+        sql =   f"""
+                    SELECT COUNT(t.id) FROM {table} t
+                    LEFT JOIN locus l ON l.id = t.{column_name}
+                    WHERE l.id IS NULL AND t.{column_name} IS NOT NULL
+                """
+
+        cursor.execute(sql)
+        data = cursor.fetchone()[0]
+        if data != 0:
+            sys.exit(f"Found {data} unlinked entries in table {table} after locus_id foreign key check\n")
+
+    db.close()
+
+def read_from_gtf(ensembl_gtf, exclude_biotypes):
+    genes_output_file = working_dir+"/ensembl_genes_grch38.txt"
+    error_log_file = working_dir+"/ensembl_genes_grch38_error.log"
+
+    gene_symbol_2_stable_id = {}
+    gene_symbol_details = {}
+    unique_gene_symbol_2_stable_id = {}
+    unique_stable_id_2_gene_symbol = {}
+
+    with gzip.open(ensembl_gtf, "rt", encoding='utf-8') as fh:
+        for line in fh:
+            if line.startswith("#"):
+                continue
+
+            data = line.strip().split("\t")
+            attribs_list = {}
+            for attrib in data[8].split(";"):
+                if attrib.strip().startswith("gene_id") or attrib.strip().startswith("gene_name") or attrib.strip().startswith("gene_biotype"):
+                    get_attrib_values = attrib.strip().replace("\"", "").split(" ")
+                    if get_attrib_values and len(get_attrib_values) == 2:
+                        attribs_list[get_attrib_values[0]] = get_attrib_values[1]
+
+            if ("gene_id" in attribs_list and attribs_list["gene_id"] and
+                "gene_name" in attribs_list and attribs_list["gene_name"] and
+                "gene_biotype" in attribs_list and attribs_list["gene_biotype"] and
+                not any(re.search(sub, attribs_list["gene_biotype"]) for sub in exclude_biotypes)):
+
+                if attribs_list["gene_name"] not in gene_symbol_2_stable_id:
```

Comment:
defaultdict can also be used here

Reply:
Same here. As the dictionary is complex I prefex to create it this way.